client/client-lib/src/api.rs

```diff
@@ -1,148 +1,615 @@
+    {
+        self.request_fed_with_strategy(
+            CurrentConsensus::new(self.all_members().one_honest()),
+            method,
+            params,
+        )
+        .await
+    }
+
+    async fn request_eventually_consistent<Ret>(
+        &self,
+        method: String,
+        params: Vec<Value>,
+    ) -> FedResult<Ret>
+    where
+        Ret: serde::de::DeserializeOwned + Eq + Debug + Clone + Send,
+    {
+        self.request_fed_with_strategy(
+            EventuallyConsistent::new(self.all_members().one_honest()),
+            method,
+            params,
+        )
+        .await
+    }
 }
 
+#[cfg_attr(target_family = "wasm", async_trait(? Send))]
+#[cfg_attr(not(target_family = "wasm"), async_trait)]
+impl<T: ?Sized> IFederationApiFed for T where T: IFederationApi {}
+
 dyn_newtype_define! {
     pub DynFederationApi(Arc<IFederationApi>)
 }
 
-impl DynFederationApi {
-    pub async fn fetch_output_outcome<T>(&self, out_point: OutPoint) -> Result<T>
+#[cfg_attr(target_family = "wasm", async_trait(? Send))]
+#[cfg_attr(not(target_family = "wasm"), async_trait)]
+pub trait GlobalFederationApi {
+    async fn get_client_config(&self) -> FedResult<ClientConfig>;
+
+    async fn submit_transaction(&self, tx: LegacyTransaction) -> FedResult<TransactionId>;
+    async fn fetch_tx_outcome(&self, txid: &TransactionId) -> FedResult<TransactionStatus>;
+
+    async fn fetch_epoch_history(
+        &self,
+        epoch: u64,
+        epoch_pk: PublicKey,
+        decoders: &ModuleDecoderRegistry,
+    ) -> FedResult<SignedEpochOutcome>;
+
+    async fn fetch_last_epoch(&self) -> FedResult<u64>;
+
+    async fn fetch_output_outcome<R>(
+        &self,
+        out_point: OutPoint,
+        decoders: &ModuleDecoderRegistry,
+    ) -> OutputOutcomeResult<R>
+    where
+        R: TryIntoOutcome + Send;
+
+    async fn await_output_outcome<R: TryIntoOutcome + Send>(
+        &self,
+        outpoint: OutPoint,
+        timeout: Duration,
+        decoders: &ModuleDecoderRegistry,
+    ) -> OutputOutcomeResult<R>;
+}
+
+#[cfg_attr(target_family = "wasm", async_trait(? Send))]
+#[cfg_attr(not(target_family = "wasm"), async_trait)]
+impl<T: ?Sized> GlobalFederationApi for T
+where
+    T: IFederationApi + Send + Sync + 'static,
+{
+    async fn get_client_config(&self) -> FedResult<ClientConfig> {
+        with_erased_no_param(|params| {
+            Box::pin(async {
+                self.request_current_consensus("/config".to_owned(), params)
+                    .await
+            })
+        })
+        .await
+    }
+
+    /// Submit a transaction for inclusion
+    async fn submit_transaction(&self, tx: LegacyTransaction) -> FedResult<TransactionId> {
+        with_erased_single_param(&SerdeTransaction::from(&tx.into_type_erased()), |params| {
+            Box::pin(async move {
+                self.request_current_consensus("/transaction".to_owned(), params)
+                    .await
+            })
+        })
+        .await
```

Comment:
Wouldn't it be easier to have a fn that generates params from serializable types instead of also making it call a closure?

Reply:
I think you're right. I started with a function that converted params to Value, and then returned Value into the desired type. But then it turned out that strategies need a return value to be already deserialized so they can do smarter things on it... but I left the thing as a closure.

Will do.